### PR TITLE
refresh auth token silently when token expires

### PR DESCRIPTION
**Pull request by kduan005**
1. **What this PR does / why we need it**:
Refresh auth token in the background when it expires.
2. **Which issue(s) this PR fixes**

        (optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged):

        Fixes #OLYMP-36711

3. **Additional information**
Screen Shot 2023-02-28 at 5 43 13 PM

4. **Special notes for your reviewer**

### CODE REVIEW

File: internal/resources/cluster/data_source_cluster.go [hunk 1]
```diff
@@ -47,6 +47,9 @@ func dataSourceTMCClusterRead(ctx context.Context, d *schema.ResourceData, m int
 				_ = schema.RemoveFromState(d, m)
 				return false, nil
 			}
+			if clienterrors.IsUnauthorizedError(err) {
```
Comment (shreyassreenivas) [inline internal/resources/cluster/data_source_cluster.go:50]: there should be refresh here

File: internal/authctx/client.go [hunk 2]
```diff
@@ -38,26 +38,28 @@ func (cfg *TanzuContext) Setup() (err error) {
...
 	cfg.TMCConnection = client.NewTestHTTPClientWithDefaultTransport()
-	return setup(cfg)
+	cfg.TMCConnection.WithHost(cfg.ServerEndpoint)
+	cfg.TMCConnection.Headers.Set("Host", cfg.ServerEndpoint)
+
+	return SetAuthCtx(cfg)
 }
 
-func setup(cfg *TanzuContext) (err error) {
+func SetAuthCtx(cfg *TanzuContext) (err error) {
```
Comment (shreyassreenivas) [inline internal/authctx/client.go:57]: why is this change needed ?

File: internal/authctx/client.go [hunk 3]
```diff
@@ -38,26 +38,28 @@ func (cfg *TanzuContext) Setup() (err error) {
 	if err != nil {
 		return err
 	}
+	cfg.TMCConnection.WithHost(cfg.ServerEndpoint)
```
Comment (shreyassreenivas) [inline internal/authctx/client.go:41]: these changes are not needed, can you please explain why this change needed ?

File: internal/resources/cluster/data_source_cluster.go (hunk 1 above)
Comment (kduan005) [reply internal/resources/cluster/data_source_cluster.go:50]: Yeah so instead of creating a new fresh util function, I'm trying to reuse the `setup()` function in `client.go` from authctx package, because essentially the setup() does two things:
- get a new auth token and set the value in `Client.Header` (which can be reused for refreshing tokens)
- set host value in `Client.Header`

And that's why I moved the second point out to where `setup()` are called, so that we don't set host value with redundancy. Feel free to let me know if a separate refresh util function is preferred.

File: internal/authctx/client.go (hunk 3 above)
Comment (kduan005) [reply internal/authctx/client.go:41]: Same as explained in below, trying to move out setting host in headers to where `setup()` is called, so that we can reuse `setup()` for refreshing auth tokens

File: internal/authctx/client.go (hunk 2 above)
Comment (kduan005) [reply internal/authctx/client.go:57]: Trying to rename it so that we can reuse it elsewhere

File: internal/resources/cluster/data_source_cluster.go (hunk 1 above)
Comment (kduan005) [reply internal/resources/cluster/data_source_cluster.go:50]: Okay so just made an update, and instead of reusing `setup()`, changed it to reusing `getUserAuthCtx()` for refreshing tokens

File: internal/resources/cluster/data_source_cluster.go [hunk 4]
```diff
@@ -48,6 +48,13 @@ func dataSourceTMCClusterRead(ctx context.Context, d *schema.ResourceData, m int
 				return false, nil
 			}
 
+			if clienterrors.IsUnauthorizedError(err) {
+				md, _ := authctx.GetUserAuthCtx(&config)
```
Comment (shreyassreenivas) [inline internal/resources/cluster/data_source_cluster.go:52]: Can we have a function defined in the terms of `refreshUserAuthCtx` and this function has these lines 

Func RefreshUserAuthCtx(){
md, _ := authctx.GetUserAuthCtx(&config)
 				for key, value := range md {
 					config.TMCConnection.Headers.Set(key, value)
 				}
}

And lets not expose GetUserAuthCtx

Comment (shreyassreenivas) [reply internal/resources/cluster/data_source_cluster.go:52]: infact you can define functional parameters to enhance the refresh usecase. 

RefreshUserAuthCtx(withRefreshOnAuthError(err))
RefreshUserAuthCtx(withForceRefresh())

This would mean that the error check is also pushed into the refresh logic

**Comment by shreyassreenivas** [comment]
Please validate if the change is working by manually testing the changes 
Test scenario
Use a attach cluster use case
set the wait period of 45minutes
Once the attach is successfully, delete the extension pod, before it communicates to tmc
This is force the cluster to wait until it is in ready state

File: internal/resources/cluster/data_source_cluster.go (hunk 4 above)
Comment (kduan005) [reply internal/resources/cluster/data_source_cluster.go]: I've updated the code according to the comment, PTAL. I'll try to get Vasundhara or Shobha's help on the test, as right now the acceptance test seems to detach the cluster prematurely before the cluster come up to the ready state. I'll post my update once I complete the test.

**Comment by kduan005** [comment]
> Please validate if the change is working by manually testing the changes Test scenario Use a attach cluster use case set the wait period of 45minutes Once the attach is successfully, delete the extension pod, before it communicates to tmc This is force the cluster to wait until it is in ready state

Screen Shot 2023-03-02 at 4 07 09 PM

I've tested with a kind cluster and added some extra logs in my local to show that the token was refreshed during the wait time, as seen in the above screenshot.

File: internal/authctx/authentication.go [hunk 5]
```diff
@@ -134,3 +134,12 @@ func getUserAuthCtx(config *TanzuContext) (map[string]string, error) {
 
 	return md, nil
 }
+
+func RefreshUserAuthCtx(config *TanzuContext, refreshCondition func(error) bool, err error) {
```
Comment (shreyassreenivas) [inline internal/authctx/authentication.go:138]: why is error needed to be returned?

File: internal/client/errors/errors.go [hunk 6]
```diff
@@ -41,6 +41,24 @@ func IsNotFoundError(err error) bool {
 	return false
 }
 
+func IsUnauthorizedError(err error) bool {
```
Comment (shreyassreenivas) [inline internal/client/errors/errors.go:44]: can you modify this function to retrun another function which will be invoked in the refreshAuthxCtx

Comment (shreyassreenivas) [reply internal/client/errors/errors.go:44]: This will reduce the error being sent as a parameter to refreshAuthxCtx

**Review by shreyassreenivas — APPROVED** [review]

**Review by Vasundharashukla — APPROVED** [review]
